samza-core/src/main/java/org/apache/samza/coordinator/metadatastore/CoordinatorStreamStore.java

```diff
@@ -0,0 +1,197 @@
+    this.coordinatorSystemStreamPartition = new SystemStreamPartition(coordinatorSystemStream, new Partition(0));
+    SystemFactory systemFactory = CoordinatorStreamUtil.getCoordinatorSystemFactory(config);
+    this.systemProducer = systemFactory.getProducer(this.coordinatorSystemStream.getSystem(), config, metricsRegistry);
+    this.systemConsumer = systemFactory.getConsumer(this.coordinatorSystemStream.getSystem(), config, metricsRegistry);
+    this.systemAdmin = systemFactory.getAdmin(this.coordinatorSystemStream.getSystem(), config);
+  }
+
+  @Override
+  public void init(SamzaContainerContext containerContext) {
+    if (isInitialized.compareAndSet(false, true)) {
+      LOG.info("Starting the coordinator stream system consumer with config: {}.", config);
+      this.containerContext = containerContext;
+      this.source = String.format("SamzaContainer-%s", this.containerContext.id);
+      registerConsumer();
+      systemConsumer.start();
+      systemProducer.register(source);
+      systemProducer.start();
+      iterator = new SystemStreamPartitionIterator(systemConsumer, coordinatorSystemStreamPartition);
+      bootstrapMessagesFromStream();
+    } else {
+      LOG.info("Store had already been initialized. Skipping.", coordinatorSystemStreamPartition);
+    }
+  }
+
+  private void registerConsumer() {
+    LOG.debug("Attempting to register system stream partition: {}", coordinatorSystemStreamPartition);
+    String streamName = coordinatorSystemStreamPartition.getStream();
+    Map<String, SystemStreamMetadata> systemStreamMetadataMap = systemAdmin.getSystemStreamMetadata(Sets.newHashSet(streamName));
+
+    SystemStreamMetadata systemStreamMetadata = systemStreamMetadataMap.get(streamName);
+    Preconditions.checkNotNull(systemStreamMetadata, String.format("System stream metadata does not exist for stream: %s.", streamName));
+
+    SystemStreamPartitionMetadata systemStreamPartitionMetadata = systemStreamMetadata.getSystemStreamPartitionMetadata().get(coordinatorSystemStreamPartition.getPartition());
+    Preconditions.checkNotNull(systemStreamPartitionMetadata, String.format("System stream partition metadata does not exist for: %s.", coordinatorSystemStreamPartition));
+
+    String startingOffset = systemStreamPartitionMetadata.getOldestOffset();
+    LOG.info("Registering system stream partition: {} with offset: {}.", coordinatorSystemStreamPartition, startingOffset);
+    systemConsumer.register(coordinatorSystemStreamPartition, startingOffset);
+  }
+
+  @Override
+  public byte[] get(byte[] key) {
+    bootstrapMessagesFromStream();
+    return bootstrappedMessages.get(key);
+  }
+
+  @Override
+  public void put(byte[] key, byte[] value) {
+    try {
+      OutgoingMessageEnvelope envelope = new OutgoingMessageEnvelope(coordinatorSystemStream, 0, key, value);
+      if (value != null) {
+        bootstrappedMessages.put(key, value);
```

Comment:
why do we need this if all the read path `get` and `all` methods bootstrap for messages prior to reading?

Reply:
Agree about the delta part. Are there possibilities of `get` or `all` being invoked concurrently during a `put`?  If no, then whatever new message that gets written (appended) to the coordinator stream should be read by subsequent `get` or `all` as part of `bootstrapMessagesFromStream()`, right?